packages/block-library/src/navigation-link/index.php

```diff
@@ -327,69 +391,42 @@ function build_variation_for_navigation_link( $entity, $kind ) {
 }
 
 /**
- * Register a variation for a post type / taxonomy for the navigation link block.
+ * Filters the registered variations for a block type.
+ * Returns the dynamically built variations for all post-types and taxonomies.
  *
- * @param array $variation Variation array from build_variation_for_navigation_link.
- * @return void
- */
-function block_core_navigation_link_register_variation( $variation ) {
-	// Directly set the variations on the registered block type
-	// because there's no server side registration for variations (see #47170).
-	$navigation_block_type = WP_Block_Type_Registry::get_instance()->get_registered( 'core/navigation-link' );
-	// If the block is not registered yet, bail early.
-	// Variation will be registered in register_block_core_navigation_link then.
-	if ( ! $navigation_block_type ) {
-		return;
-	}
-
-	$navigation_block_type->variations = array_merge(
-		$navigation_block_type->variations,
-		array( $variation )
-	);
-}
-
-/**
- * Unregister a variation for a post type / taxonomy for the navigation link block.
+ * @since 6.5.0
  *
- * @param string $name Name of the post type / taxonomy (which was used as variation name).
- * @return void
+ * @param array         $variations Array of registered variations for a block type.
+ * @param WP_Block_Type $block_type The full block type object.
  */
-function block_core_navigation_link_unregister_variation( $name ) {
-	// Directly get the variations from the registered block type
-	// because there's no server side (un)registration for variations (see #47170).
-	$navigation_block_type = WP_Block_Type_Registry::get_instance()->get_registered( 'core/navigation-link' );
-	// If the block is not registered (yet), there's no need to remove a variation.
-	if ( ! $navigation_block_type || empty( $navigation_block_type->variations ) ) {
-		return;
+function block_core_navigation_link_filter_variations( $variations, $block_type ) {
+	if ( 'core/navigation-link' !== $block_type->name ) {
+		return $variations;
 	}
-	$variations = $navigation_block_type->variations;
-	// Search for the variation and remove it from the array.
-	foreach ( $variations as $i => $variation ) {
-		if ( $variation['name'] === $name ) {
-			unset( $variations[ $i ] );
-			break;
-		}
-	}
-	// Reindex array after removing one variation.
-	$navigation_block_type->variations = array_values( $variations );
+
+	$generated_variations = block_core_navigation_link_build_variations();
+	// TODO: Mabe check if there are already variations for this post type/taxonomy?
+	return array_merge( $variations, $generated_variations );
```

Comment:
@gaambo What happens if there are already variations registered? Do we double up? We should be able to de-duplicate here right?

Reply:
My TODO comment was meant for the case, that a developer of a custom post type/taxonomy already added a variation for the navigation-link block for their post type/taxonomy and then there would be  two variations. I think this case is unlikely, but still. 
The array merge does not handle doubles in any way(since it doesn't handle nested arrays afaik).  The only way left is to check for each automatically generated variation, if there is already a variation with this name - since the array is not indexed by variation name/keys, this means searching/looping through the array a few times. Is that worth it? 
I can't see any code in `WP_Block_Type` handling multiple variations with the same key, don't know about the editor side.

Reply:
Let's leave "as is" and remove the TODO comment. I think the scenario is unlikely and if it becomes a problem it can be patched during the remainder of the release.